File: csharp/affine-cipher/AffineCipher.cs
using System;
using System.Linq;

public static class AffineCipher
{
    const int m = 26;

    public static string Encode(string plainText, int a, int b)
    {
        if(!AreCoPrime(a, m))
            throw new ArgumentException();

        int algo(char c) => ((c - 'a') * a + b) % m;
        char asChar(int cipher) => (char)(cipher + 'a');

        var cleaned = plainText.ToLower().Where(char.IsLetterOrDigit);
        var encoded = cleaned.Select(c => char.IsDigit(c) ? c : asChar(algo(c))).ToArray();
        var chunked = Enumerable.Range(0, encoded.Length / 5 + 1)
            .Select(i => new string(encoded.Skip(i * 5).Take(5).ToArray()));

        return string.Join(' ', chunked.ToArray()).Trim();
    }

    public static string Decode(string cipheredText, int a, int b)
    {
        if(!AreCoPrime(a, m))
            throw new ArgumentException();

        var modInverse = Enumerable.Range(1, m).First(i => i * a % m == 1);
        int algo(char c) => (modInverse * ((c - 'a') - b)) % m;
        char asChar(int cipher) => (char)((cipher < 0 ? 26 + cipher : cipher) + 'a');

        var cleaned = cipheredText.ToLower().Where(char.IsLetterOrDigit);
        var decoded = cleaned.Select(c => char.IsDigit(c) ? c : asChar(algo(c))).ToArray();

        return new string(decoded);
    }

    private static bool AreCoPrime(int value1, int value2)
    {
        while (value1 != 0 && value2 != 0)
        {
            if (value1 > value2)
                value1 %= value2;
            else
                value2 %= value1;
        }
        return Math.Max(value1, value2) == 1;
    }
}

File: fsharp/affine-cipher/AffineCipher.fs
module AffineCipher

open System

let m = 26

let coprime a b =
    let mutable p1, p2 = a, b
    while p1 <> 0 && p2 <> 0 do
        if p1 > p2 then p1 <- p1 % p2
        else p2 <- p2 % p1
    (max p1 p2) = 1

let clean (s: string) = 
    s.ToLower() |> Seq.filter Char.IsLetterOrDigit

let decode a b cipheredText = 
    if not (coprime a m) then
        invalidArg "a" (sprintf "%i is not coprime with %i" a m)

    let modInverse = [1..m] |> List.find (fun i -> i * a % m = 1) 
    let decode (c: char) = 
        if Char.IsDigit c then c
        else
            let v = (modInverse * ((int c - int 'a') - b)) % m
            let vc = if v < 0 then v + m else v
            char (vc + int 'a')

    clean cipheredText
    |> Seq.map decode 
    |> Seq.toArray
    |> String

let encode a b plainText = 
    if not (coprime a m) then
        invalidArg "a" (sprintf "%i is not coprime with %i" a m)
    
    let encode (c: char) =
        if Char.IsDigit c then c
        else
            let v = ((int c - int 'a') * a + b) % m
            char (v + int 'a')

    clean plainText
    |> Seq.map encode 
    |> Seq.toArray
    |> Array.chunkBySize 5 
    |> Array.map String 
    |> String.concat " "
